cluster_management/src/main/java/com/pinterest/rocksplicator/shardmap/JsonShardMap.java

```diff
@@ -0,0 +1,158 @@
+import org.json.simple.JSONObject;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+class JsonShardMap implements ShardMap {
+
+  private final Map<String, ResourceMap> resourceMap;
+
+  JsonShardMap(JSONObject shardMapObj) {
+    resourceMap = new HashMap<>(shardMapObj.size());
+
+    for (Object resourceNameObj : shardMapObj.keySet()) {
+      String resourceName = (String) resourceNameObj;
+      JSONObject resourceObj = (JSONObject) shardMapObj.get(resourceNameObj);
+      if (resourceObj != null) {
+        resourceMap.put(resourceName, new JsonResourceMapImpl(resourceName, resourceObj));
+      }
+    }
+  }
+
+  @Override
+  public Set<String> getResources() {
+    return resourceMap.keySet();
+  }
+
+  @Override
+  public ResourceMap getResourceMap(String resource) {
+    return resourceMap.get(resource);
+  }
+
+
+  private class JsonResourceMapImpl implements ResourceMap {
+
+    private final String resourceName;
+    private final int numShards;
+
+    private Map<Partition, List<Replica>> replicasByPartition;
+    private Map<Instance, List<Replica>> replicasByInstance;
+    private Set<Partition> allPartitions;
+
+    JsonResourceMapImpl(String resourceName, JSONObject resourceMapObj) {
+      this.resourceName = resourceName;
+      this.replicasByPartition = new HashMap<>();
+      this.replicasByInstance = new HashMap();
+
+      Object numShardsObj = resourceMapObj.get("num_shards");
+      if (numShardsObj.getClass().equals(Integer.class)) {
+        this.numShards = (Integer) numShardsObj;
+      } else if (numShardsObj.getClass().equals(Long.class)) {
+        this.numShards = ((Long) numShardsObj).intValue();
+      } else {
+        throw new RuntimeException("Illegal format for num_shards for resource:" + resourceName + ", json string is " + resourceMapObj.toJSONString());
+      }
+
+
+
+      ImmutableSet.Builder<Partition> setBuilder = ImmutableSet.builder();
+      for (int partitionId = 0; partitionId < numShards; ++partitionId) {
+        setBuilder.add(new Partition(resourceName, partitionId));
+      }
+      this.allPartitions = setBuilder.build();
+
+      // Now for all instance info, construct Replicas information.
+      for (Object key : resourceMapObj.keySet()) {
+        if ("num_shards".equalsIgnoreCase((String) key)) {
+          continue;
+        }
+        String encodedInstanceStr = (String) key;
+        Instance instance = new Instance(encodedInstanceStr);
+
+        JSONArray partitionsOnInstance = (JSONArray) resourceMapObj.get(key);
+        for (Object encodedPartitionObj : partitionsOnInstance) {
+          String encodedPartitionStr = (String) encodedPartitionObj;
+          String parts[] = encodedPartitionStr.split(":");
+          Partition partition = new Partition(resourceName, parts[0]);
+          ReplicaState replicaState = ReplicaState.ONLINE;
```

Comment:
why default as online?

Reply:
If the partition info doesn't contain ':' then it cannot be Master/Slave or Leader/Follower model. Since this entry exists in the JSON ShardMap, it must be ONLINE. However, this is for completeness, in representation of a Partition replica state. We are only interested in serving partitions. And shardmap only contains serving partitions. This is also what client's assume.